ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java

```diff
@@ -85,28 +88,63 @@ void fetchAndRunCommands() {
         .forEach(c -> executeStatement(c.getCommand().get(), c.getCommandId(), c.getStatus()));
   }
 
+
+
   /**
    * Read and execute all commands on the command topic, starting at the earliest offset.
    * @throws Exception TODO: Refine this.
    */
   public void processPriorCommands() {
     final RestoreCommands restoreCommands = commandStore.getRestoreCommands();
-    statementExecutor.handleRestoration(restoreCommands);
+    restoreCommands.forEach(
+        (commandId, command, terminatedQueries, wasDropped) -> {
+          executeStatementWithRetries(
+              () -> statementExecutor.handleStatementWithTerminatedQueries(
+                  command,
+                  commandId,
+                  Optional.empty(),
+                  terminatedQueries,
+                  wasDropped
+              )
+          );
+        }
+    );
   }
 
   private void executeStatement(final Command command,
                                 final CommandId commandId,
                                 final Optional<QueuedCommandStatus> status) {
     log.info("Executing statement: " + command.getStatement());
-    try {
-      statementExecutor.handleStatement(command, commandId, status);
-    } catch (final WakeupException wue) {
-      throw wue;
-    } catch (final Exception exception) {
-      final StringWriter stringWriter = new StringWriter();
-      final PrintWriter printWriter = new PrintWriter(stringWriter);
-      exception.printStackTrace(printWriter);
-      log.error("Exception encountered during poll-parse-execute loop: " + stringWriter.toString());
+    executeStatementWithRetries(
+        () -> statementExecutor.handleStatement(command, commandId, status)
+    );
+  }
+
+  private void executeStatementWithRetries(final Runnable runnable) {
+    RuntimeException finalException = null;
+    for (int i = 0; i < maxRetries && !closed; i++) {
+      try {
+        runnable.run();
```

Comment:
Do we need to differentiate between exceptions that may succeed if retried, e.g. things that interact with a remote service that is currently unavailable, verse something that will always fail, e.g. a bad command?

Reply:
I think there's a few classes of errors:

- Input errors - something passed to us (e.g. the statement) was messed up.
- External errors - some system we depend on has failed.
- Internal errors - something went wrong inside ksql (e.g. a bug caused some state to be 
corrupted which caused some validation to fail or some unexpected condition to arise)

Input errors should just be returned to the user. For the other errors we should block until the error condition resolves. At some point (arguably immediately for internal errors) it probably makes sense to abort the process too, though for now its at least better to keep retrying in-process than to just move on and potentially create an inconsistent metastore.

The trick is of course to differentiate the error types. This patch interprets KsqlException as an input error (StatementExecutor catches it and fails the request). Looking through the code this is the case almost everywhere. There's a couple places where we do throw KsqlException on non-user errors. I think that clean-up can come in a follow-up PR. But even for those cases the behaviour is no worse than it used to be.

Reply:
Agree with renaming. I've filed https://github.com/confluentinc/ksql/issues/2038 to track the follow-up work.